File: Java/P4/irrgarten/irrgarten/src/irrgarten/Player.java

package irrgarten;
import java.util.ArrayList;
import java.util.Iterator;

/**
 * @brief   Funcionalidades de los jugadores
 * @author  Isabel Morro Tabares
 */
public class Player extends LabyrinthCharacter {
   
    // Declaración de atributos
    
    private static final int MAX_WEAPONS = 2;
    private static final int MAX_SHIELDS = 3;
    private static final int INITIAL_HEALTH = 10;
    private static final int HITS2LOSE = 3;
    
    private char number;
    private int consecutiveHits;
    
    private ArrayList<Weapon> weapons;
    private ArrayList<Shield> shields;
    
    private ShieldCardDeck shieldCardDeck;
    private WeaponCardDeck weaponCardDeck;
    
    // Constructor
    
    public Player (char number, float intelligence, float strength){
        super ("Player #" + number, intelligence, strength);
        this.number = number;
        
        weapons = new ArrayList<>();
        shields = new ArrayList<>();
        shieldCardDeck = new ShieldCardDeck();
        weaponCardDeck = new WeaponCardDeck();

        setHealth(INITIAL_HEALTH);
        consecutiveHits = 0;
        setPos(-1, -1);
    }
    
    public Player (Player other){
        super (other);
        this.number = other.getNumber();
    }
    
    // Métodos privados
    
    private void receiveWeapon(Weapon w){
        
        int tam_ini = weapons.size();
        
        for (int i=tam_ini-1; i >= 0; i--){
            if (weapons.get(i).discard())
                weapons.remove(i);
        }
        
        if (weapons.size() < MAX_WEAPONS) weapons.add(w);
        
    }
    
    private void receiveShield(Shield s){
        
        int tam_ini = shields.size();
        
        for (int i=tam_ini-1; i >= 0; i--){
            if (shields.get(i).discard())
                shields.remove(i);
        }
       
        if (shields.size() < MAX_SHIELDS) shields.add(s);
    }
    
    private Weapon newWeapon() {
        return weaponCardDeck.nextCard();
    }
    
    private Shield newShield(){
        return shieldCardDeck.nextCard();
    }
    
    protected float sumWeapons(){
        float suma = 0f;
        if (weapons != null){
            for (int i=0; i < weapons.size(); i++)
                suma += weapons.get(i).attack();
        }
        return suma;
    }
    
    protected float sumShields(){
        
        float suma = 0f;
        if (shields != null){
            for(int i=0; i < shields.size(); i++)
                suma += shields.get(i).protect();
        }
        return suma;
    }
    
    protected float defensiveEnergy(){
        return sumShields() + getIntelligence();
    }
    
    private boolean manageHit (float receivedAttack){
        
        float defense = defensiveEnergy();
        boolean lose;
        
        if (defense < receivedAttack){
            gotWounded();
            incConsecutiveHits();
        }
        else resetHits();
        
        if ( consecutiveHits == HITS2LOSE || dead()){
            resetHits();
            lose = true;
        }
        else lose = false;
        
        return lose;
    }
    
    private void resetHits(){
        consecutiveHits = 0;
    }

    private void incConsecutiveHits(){
        consecutiveHits++;
    }
    
    // Métodos públicos
    
    public void resurrect (){
        weapons.clear(); 
        shields.clear();
        setHealth (INITIAL_HEALTH);
        resetHits();
    }
    

    public char getNumber(){
        return number;
    }
    
    public Directions move(Directions direction, ArrayList<Directions> validMoves){
        
        int size = validMoves.size();
        boolean contained = validMoves.contains(direction);
        Directions result;
        
        if (size>0 && !contained)
            result = validMoves.get(0);
        else
            result = direction;
        
        return direction;
    }
    
    @Override
    public float attack(){
        return sumWeapons() + getStrength();
    }
    
    
    @Override
    public boolean defend (float receivedAttack){
        return manageHit(receivedAttack);
    }
    
    public void receiveReward(){
        
        int wReward = Dice.weaponsReward();
        int sReward = Dice.shieldsReward();
        
        for (int i=0; i < wReward; i++){
            Weapon wnew = newWeapon();
            receiveWeapon(wnew);
        }
        
        for (int i=0; i < sReward; i++){
            Shield snew = newShield();
            receiveShield(snew);
        }
        
        int extraHealth = Dice.healthReward();
        setHealth(getHealth() + extraHealth);
    }
    
    @Override
    public String toString(){
        String cadena = super.toString();
        cadena += " Numero: " + number + ","
                + " Impactos consecutivos: " 
                + consecutiveHits + "\n" 
                + " Armas: ";
        
        if (weapons != null){
            if (!weapons.isEmpty()){
                for (Weapon wi : weapons) cadena += wi.toString() + " ";
            }
        }
        else cadena += "No dispone de armas";
        cadena += "\n";
        cadena += " Shields: ";
        if (shields != null){
            if (!shields.isEmpty()){
                for (Shield si : shields) cadena += si.toString() + " ";
            }
        }
        else cadena += "No dispone de escudos";
        cadena += "]";
        
        return cadena;
    }
   
}

File: Ruby/P4/Irrgarten/player.rb
#encoding:utf-8

=begin

    Autor: Isabel Morro Tabares 

    Funcionalidades de los jugadores

=end

require_relative 'dice'
require_relative 'directions'
require_relative 'weapon'
require_relative 'shield'
require_relative 'labyrinth_character'

module Irrgarten
    class Player < LabyrinthCharacter
    
    private 

        @@MAX_WEAPONS = 2
        @@MAX_SHIELDS = 3
        @@INITIAL_HEALTH = 10
        @@HITS_2_LOSE = 3

        # Métodos privados

        def receive_weapon(w)

            @weapons.each do |wi|
                discard = wi.discard()
                if (discard)
                    @weapons.delete(wi)
                end
            end

            size = @weapons.size

            if (size < @@MAX_WEAPONS)
                @weapons << w  
            end

        end

        def receive_shield(s)
        
            @shields.each do |si|
                discard = si.discard()
                if (discard)
                    @shields.delete(si)
                end
            end

            size = @shields.size

            if (size < @@MAX_SHIELDS)
                @shields << s 
            end
        
        end

        def new_weapon
            power = Dice.weapon_power
            uses = Dice.uses_left
            w = Weapon.new(power, uses)
            w
        end

        def new_shield            
            protection = Dice.shield_power
            uses = Dice.uses_left
            s = Shield.new(protection, uses)
            s
        end

        def sum_weapons
            suma = 0
                @weapons.each do |w|
                    suma += w.attack
                end
            suma
        end

        def sum_shields
            suma = 0
                @shields.each do |s|
                    suma += s.protect
                end
            suma
        end

        def defensive_energy
            sum_shields() + @intelligence
        end

        def manage_hit(receive_attack)

            defense = defensive_energy()

            if (defense < receive_attack)
                got_wounded()
                inc_consecutive_hits()
            else
                reset_hits()
            end

            if (@consecutive_hits == @@HITS_2_LOSE || dead())
                reset_hits()
                lose = true
            else
                lose = false 
            end

            lose
        end

        def reset_hits
            @consecutive_hits = 0
        end

        def inc_consecutive_hits
            @consecutive_hits += 1
        end

    public 
    
        # Métodos públicos

        def initialize (number, intelligence, strength)
            super("Player #" + number, intelligence, strength)
            @number = number
            @consecutive_hits = 0
            @weapons = Array.new
            @shields = Array.new
            set_health(@@INITIAL_HEALTH)
        end

        def copia(other)
            super
        end

        def resurrect
            @weapons.clear 
            @shields.clear 
            @health = @@INITIAL_HEALTH
            reset_hits()
        end

        def number
            @number
        end

        def move (direction, valid_moves)

            size = valid_moves.length
            contained = valid_moves.include?(direction)

            if (size > 0 && !contained)
                first_element = valid_moves[0]
                resultado = first_element
            else
                resultado = direction
            end

            return resultado

        end

        def attack
            sum_weapons + strength
        end

        def defend (receive_attack)
            manage_hit(receive_attack)
        end

        def receive_reward

            w_reward = Dice.weapons_reward()
            s_reward = Dice.shields_reward()

            for i in (0..w_reward)
                w_new = new_weapon()
                receive_weapon(w_new)
            end

            for i in (0..s_reward)
                s_new = new_shield()
                receive_shield(s_new)
            end

            extra_health = Dice.health_reward()
            @health += extra_health

        end

        # To string
        def to_s
            cadena = super
            cadena  += " Numero: " + @number + ","
            cadena  += " Impactos consecutivos: " + @consecutive_hits.to_s + "\n"
            cadena  += " Armas: "
         
            if (!@weapons.empty?)
                @weapons.each do |wi|
                    cadena += wi.to_s + " "
                end
            else
                cadena += "No dispone de armas"
            end
            cadena += "\n"
            cadena += " Shields: "
            if (!@shields.empty?)
                @shields.each do |si|
                    cadena += si.to_s + " "
                end
            else 
                cadena += "No dispone de escudos"
            end
            cadena += "]\n"
            
            cadena
        
        end

    
    end
end
